### PR TITLE
add integration tests for rank, dense_rank, fix last_value evaluation with rank

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

add integration tests for rank, dense_rank, fix last_value evaluation with rank

Closes follow ups on #555 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/tests/sql.rs [hunk 1]
```diff
@@ -952,15 +952,8 @@ async fn csv_query_window_with_order_by() -> Result<()> {
     let actual = execute(&mut ctx, sql).await;
     let expected = vec![
```
Comment (Jimexist) [inline datafusion/tests/sql.rs:953]:
```
-[ RECORD 1 ]-----------------------
c9          | 28774375
sum         | 61035129
avg         | 61035129.000000000000
count       | 1
max         | 61035129
min         | 61035129
first_value | 61035129
last_value  | 61035129
nth_value   | ¤
-[ RECORD 2 ]-----------------------
c9          | 63044568
sum         | -47938237
avg         | -23969118.500000000000
count       | 2
max         | 61035129
min         | -108973366
first_value | 61035129
last_value  | -108973366
nth_value   | -108973366
-[ RECORD 3 ]-----------------------
c9          | 141047417
sum         | 575165281
avg         | 191721760.33333333
count       | 3
max         | 623103518
min         | -108973366
first_value | 61035129
last_value  | 623103518
nth_value   | -108973366
-[ RECORD 4 ]-----------------------
c9          | 141680161
sum         | -1352462829
avg         | -338115707.25000000
count       | 4
max         | 623103518
min         | -1927628110
first_value | 61035129
last_value  | -1927628110
nth_value   | -108973366
-[ RECORD 5 ]-----------------------
c9          | 145294611
sum         | -3251637940
avg         | -650327588.00000000
count       | 5
max         | 623103518
min         | -1927628110
first_value | 61035129
last_value  | -1899175111
nth_value   | -108973366
```

**Comment by Jimexist** [comment]
@alamb and @Dandandan this pull request is ready now

File: datafusion/src/physical_plan/expressions/nth_value.rs [hunk 2]
```diff
@@ -138,21 +140,56 @@ pub(crate) struct NthValueEvaluator {
...
+                Ok(value.to_array_of_size(num_rows))
+            }
+            NthValueKind::Last => {
+                // because the default window frame is between unbounded preceding and current
+                // row with peer evaluation, hence the last rows expands until the end of the peers
+                let values = ranks_in_partition
+                    .iter()
+                    .map(|range| {
+                        let len = range.end - range.start;
+                        let value = ScalarValue::try_from_array(arr, range.end - 1)?;
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/nth_value.rs:170]: this is very cool

File: datafusion/src/physical_plan/expressions/nth_value.rs [hunk 3]
```diff
@@ -138,21 +140,56 @@ pub(crate) struct NthValueEvaluator {
...
+                if index >= num_rows {
+                    Ok(new_null_array(arr.data_type(), num_rows))
+                } else {
+                    let value =
+                        ScalarValue::try_from_array(arr, partition.start + index)?;
+                    let arr = value.to_array_of_size(num_rows);
+                    // because the default window frame is between unbounded preceding and current
+                    // row, hence the shift because for values with indices < index they should be
+                    // null. This changes when window frames other than default is implemented
+                    shift(arr.as_ref(), index as i64).map_err(DataFusionError::ArrowError)
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/nth_value.rs:189]: 👍

**Review by alamb — APPROVED** [review]
This looks great to me. Nice work @Jimexist

**Pull request merged**